src/genetic_algorithm.py

```diff
@@ -0,0 +1,125 @@
+import numpy as np
+import matplotlib.pyplot as plt
+import matplotlib.animation as animation
+
+
+# Problem settings
+def sum_equation(x):
+    """Example equation: the sum of squares."""
+    return sum([i**2 for i in x])
+
+
+# Genetic Algorithm
+class GeneticAlgorithm:
+    def __init__(
+        self, pop_size, dimensions, mutation_rate, crossover_rate, max_generations
+    ):
+        self.pop_size = pop_size
+        self.dimensions = dimensions
+        self.mutation_rate = mutation_rate
+        self.crossover_rate = crossover_rate
+        self.max_generations = max_generations
+        self.population = np.random.randn(pop_size, dimensions)
+        self.best_solution = None
+        self.best_fitness = float("inf")
+        self.fitness_history = []
+
+    def fitness(self, solution):
+        return sum_equation(solution)
+
+    def select_parents(self):
+        """Tournament selection"""
+        parents = []
+        for _ in range(self.pop_size):
+            i, j = np.random.randint(0, self.pop_size, 2)
+            if self.fitness(self.population[i]) < self.fitness(self.population[j]):
+                parents.append(self.population[i])
+            else:
+                parents.append(self.population[j])
+        return np.array(parents)
+
+    def crossover(self, parent1, parent2):
+        if np.random.rand() < self.crossover_rate:
+            crossover_point = np.random.randint(1, self.dimensions)
+            child1 = np.concatenate(
+                [parent1[:crossover_point], parent2[crossover_point:]]
+            )
+            child2 = np.concatenate(
+                [parent2[:crossover_point], parent1[crossover_point:]]
+            )
+            return child1, child2
+        else:
+            return parent1, parent2
```

Comment:
These will be difficult to test. You will need to either (a) mock the random library or (b) just pass the random values to the function

Reply:
Updated the function to use rand (float or None) for Optional random value for testing purposes. If not provided, it will use `np.random.rand()`.